### PR TITLE
Caffe 2: Reshape Op upgrade

### PR DESCRIPTION
**Pull request by drnikolaev**
This is follow up on #13945 where we had to turn off some TRT tests because some ops were not ready to accept ONNX opset 9+ models. This PR fixes Reshape.

### CODE REVIEW

**Comment by ezyang** [comment]
@houseroad  mind taking a look at this?

**Comment by houseroad** [comment]
Sure, I will take a look.

**Comment by drnikolaev** [comment]
@ezyang @houseroad there is one failure I'm still working on...

File: caffe2/operators/reshape_op.h [hunk 1]
```diff
@@ -46,14 +46,20 @@ class ReshapeOp : public Operator<Context> {
...
 
-      const T* shape_data = shape.template data<T>();
-
-      // Bit awkward, but needed so works on both CPU and CUDA contexts
-      std::vector<T> tmpv(shape.numel());
-      if (shape.numel() > 0) {
-        context_.CopyBytesToCPU(
-            shape.numel() * sizeof(T), shape_data, &tmpv[0]);
-        actual_new_shape.assign(tmpv.begin(), tmpv.begin() + shape.numel());
+      if (shape.dtype() == TypeMeta::Make<int64_t>() ||
```
Comment (houseroad) [inline caffe2/operators/reshape_op.h]: Why do we need to change the implementation of Reshape in Caffe2?

File: caffe2/python/operator_test/reshape_ops_test.py [hunk 2]
```diff
@@ -16,16 +16,16 @@ def test_lengths_to_shape_ops(self):
         workspace.RunOperatorOnce(core.CreateOperator(
             'LengthsToShape', ['l'], ['s']))
         workspace.FeedBlob('res', np.array([3, 200], dtype=np.int32))
-        assert ((workspace.FetchBlob('s') == workspace.FetchBlob('res')).all())
+        assert (workspace.FetchBlob('s') == workspace.FetchBlob('res')).all()
```
Comment (houseroad) [inline caffe2/python/operator_test/reshape_ops_test.py]: Can we replace python assert with numpy assert_array_equal?

**Review by houseroad — CHANGES_REQUESTED** [review]
Sorry, I am confused here, why do we need to change Caffe2 operator implementation here?

File: caffe2/operators/reshape_op.h (hunk 1 above)
Comment (drnikolaev) [reply caffe2/operators/reshape_op.h]: Well, this is not changing but rather adding one more use case when `Input(1)` appears to be float, thus it fails here. It happens with ONNX zoo models of opset 5+. So, as you can see, we have if/else fork here. However, I might be wrong and should reject such cases. Indeed, when I get `shape=[1.4678, 0.43289]` what exactly this means?

File: caffe2/python/operator_test/reshape_ops_test.py (hunk 2 above)
Comment (drnikolaev) [reply caffe2/python/operator_test/reshape_ops_test.py]: Sure, will do after holidays :)

**Comment by drnikolaev** [comment]
@houseroad @ezyang could you please have a look again?

**Comment by houseroad** [comment]
@drnikolaev sure, i will take a look.

File: caffe2/python/trt/test_trt.py [hunk 3]
```diff
@@ -131,46 +131,48 @@ def _test_onnx_importer(self, model_name, data_input_index,
         ws = Workspace()
         with core.DeviceScope(device_option):
             ws.FeedBlob(op_inputs[data_input_index], data)
+            if opset_version >= 5:
+                ws.FeedBlob("data_0", data)
```
Comment (houseroad) [inline caffe2/python/trt/test_trt.py]: Looks like we are hard coded the input name. If so, please leave some comments to explain it.

File: caffe2/python/onnx/backend.py [hunk 4]
```diff
@@ -31,7 +31,7 @@
 import caffe2.python.utils
 import numpy as np
 import onnx
-from onnx import checker, GraphProto, TensorProto, AttributeProto, ModelProto
+from onnx import checker, GraphProto, TensorProto, AttributeProto, ModelProto, mapping
```
Comment (houseroad) [inline caffe2/python/onnx/backend.py]: Is `mapping` necessary? I didn't see it was used anywhere.

**Review by houseroad — COMMENTED** [review]
Left some inline comments

File: caffe2/python/trt/test_trt.py (hunk 3 above)
Comment (drnikolaev) [reply caffe2/python/trt/test_trt.py]: Done. Thanks!

File: caffe2/python/onnx/backend.py (hunk 4 above)
Comment (drnikolaev) [reply caffe2/python/onnx/backend.py]: Ouch

Comment (drnikolaev) [reply caffe2/python/onnx/backend.py]: Removed

**Review by houseroad — APPROVED** [review]
Looks good, thanks!

**Pull request closed without merging**